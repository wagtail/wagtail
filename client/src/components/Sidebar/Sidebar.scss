@mixin sidebar-toggle() {
    @include transition(background-color $menu-transition-duration ease);

    position: absolute;
    top: 12px;
    left: 12px;
    color: $color-white;
    width: 35px;
    height: 35px;
    background: transparent;
    border: 0;
    place-items: center;
    padding: 0;
    border-radius: 50%;

    &:hover {
        background-color: #3a3a3a;
        color: #ccc;
    }

    svg {
        width: 15px;
        height: 16px;
    }
}

.sidebar {
    position: fixed;
    left: 0;
    width: $menu-width;
    float: left;
    display: flex;
    flex-direction: column;
    height: 100%;
    background: $nav-grey-3;
    z-index: 3;

    @include transition(width $menu-transition-duration ease, left $menu-transition-duration ease);

    &--slim {
        width: $menu-width-slim;
    }

    // The sidebar can move completely off-screen in mobile mode for extra room
    &--hidden {
        left: -$menu-width;
    }

    &__inner {
        height: 100%;
        background: $nav-grey-3;
        // On medium, make it possible for the nav links to scroll.
        display: flex;
        flex-flow: column nowrap;
    }

    &__collapse-toggle {
        @include sidebar-toggle;
        display: grid;
    }

    // When in mobile mode, hide the collapse-toggle and show the nav-toggle (which is defined in the .sidebar-nav-toggle class below)
    &--mobile &__collapse-toggle {
        display: none;
    }

    // This element should cover all the area beneath the collapse toggle
    // It's only used to attach mouse enter/exit event handlers to control peeking
    &__peek-hover-area {
        margin-top: 35px;  // Match height of __collapse-toggle
        display: grid;
        grid-template-columns: 1fr;
        overflow-y: scroll;
    }
}

// This is a separate component as it needs to display in the header
.sidebar-nav-toggle {
    @include sidebar-toggle;
    display: none;  // Nav toggle is for mobile only
    z-index: 5;

    &:hover {
        background: var(--color-primary-dark);
        color: #fff;
    }

    &--mobile {
        display: grid;
    }

    &--open {
        position: fixed;

        &:hover {
            background-color: #3a3a3a;
            color: #ccc;
        }
    }
}


@import 'SidebarPanel';
@import 'menu/MenuItem';
@import 'menu/SubMenuItem';
@import 'modules/CustomBranding';
@import 'modules/MainMenu';
@import 'modules/Search';
@import 'modules/WagtailBranding';
