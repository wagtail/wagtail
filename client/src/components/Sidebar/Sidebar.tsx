import * as React from 'react';

import { gettext } from '../../utils/gettext';
import Icon from '../Icon/Icon';

// Please keep in sync with $menu-transition-duration variable in `client/scss/settings/_variables.scss`
export const SIDEBAR_TRANSITION_DURATION = 150;

export interface ModuleRenderContext {
  key: number;
  slim: boolean;
  expandingOrCollapsing: boolean;
  onHideMobile: () => void;
  onSearchClick: () => void;
  currentPath: string;
  navigate(url: string): Promise<void>;
}

export interface ModuleDefinition {
  render(context: ModuleRenderContext): React.ReactFragment;
}

export interface SidebarProps {
  modules: ModuleDefinition[];
  currentPath: string;
  collapsedOnLoad: boolean;
  navigate(url: string): Promise<void>;
  onExpandCollapse?(collapsed: boolean);
}

export const Sidebar: React.FunctionComponent<SidebarProps> = ({
  modules,
  currentPath,
  collapsedOnLoad = false,
  navigate,
  onExpandCollapse,
}) => {
  // 'collapsed' is a persistent state that is controlled by the arrow icon at the top
  // It records the user's general preference for a collapsed/uncollapsed menu
  // This is just a hint though, and we may still collapse the menu if the screen is too small
  const [collapsed, setCollapsed] = React.useState(collapsedOnLoad);
  const mobileNavToggleRef = React.useRef<HTMLButtonElement>(null);

  // Call onExpandCollapse(true) if menu is initialised in collapsed state
  React.useEffect(() => {
    if (collapsed && onExpandCollapse) {
      onExpandCollapse(true);
    }
  }, []);

  // 'visibleOnMobile' indicates whether the sidebar is currently visible on mobile
  // On mobile, the sidebar is completely hidden by default and must be opened manually
  const [visibleOnMobile, setVisibleOnMobile] = React.useState(false);
  // 'closedOnMobile' is used to set the menu to display none so it can no longer be interacted with by keyboard when its hidden
  const [closedOnMobile, setClosedOnMobile] = React.useState(true);

  // Tracks whether the screen is below 800 pixels. In this state, the menu is completely hidden.
  // State is used here in case the user changes their browser size
  const checkWindowSizeIsMobile = () => window.innerWidth < 800;
  const [isMobile, setIsMobile] = React.useState(checkWindowSizeIsMobile());
  React.useEffect(() => {
    function handleResize() {
      if (checkWindowSizeIsMobile()) {
        setIsMobile(true);
        return null;
      }
      setIsMobile(false);

      // Close the menu and animate out as this state is not used in desktop
      setVisibleOnMobile(false);
      // wait for animation to finish then hide menu from screen readers as well.
      return setTimeout(() => {
        setClosedOnMobile(true);
      }, SIDEBAR_TRANSITION_DURATION);
    }

    window.addEventListener('resize', handleResize);
    const closeTimeout = handleResize();
    return () => {
      window.removeEventListener('resize', handleResize);
      if (closeTimeout) {
        clearTimeout(closeTimeout);
      }
    };
  }, []);

  // Whether or not to display the menu with slim layout.
  const slim = collapsed && !isMobile;

  // 'expandingOrCollapsing' is set to true whilst the menu is transitioning between slim and expanded layouts
  const [expandingOrCollapsing, setExpandingOrCollapsing] =
    React.useState(false);

  React.useEffect(() => {
    setExpandingOrCollapsing(true);
    const finishTimeout = setTimeout(() => {
      setExpandingOrCollapsing(false);
    }, SIDEBAR_TRANSITION_DURATION);

    return () => {
      clearTimeout(finishTimeout);
    };
  }, [slim]);

  const onClickCollapseToggle = () => {
    setCollapsed(!collapsed);

    if (onExpandCollapse) {
      onExpandCollapse(!collapsed);
    }
  };

  const onClickOpenCloseToggle = () => {
    setVisibleOnMobile(!visibleOnMobile);
    setExpandingOrCollapsing(true);

    const finishTimeout = setTimeout(() => {
      setExpandingOrCollapsing(false);
      setClosedOnMobile(!closedOnMobile);
    }, SIDEBAR_TRANSITION_DURATION);
    return () => {
      clearTimeout(finishTimeout);
    };
  };

  const [focused, setFocused] = React.useState(false);

  const onBlurHandler = () => {
    if (focused) {
      setFocused(false);
      setCollapsed(true);
    }
  };

  const onFocusHandler = () => {
    if (focused) {
      setCollapsed(false);
      setFocused(true);
    }
  };

  const onSearchClick = () => {
    if (slim) {
      onClickCollapseToggle();
    }
  };

  React.useEffect(() => {
    // wait for animation to finish then hide menu from screen readers as well.
    const finishHidingMenu = setTimeout(() => {
      if (!visibleOnMobile) {
        setClosedOnMobile(true);
      }
    }, SIDEBAR_TRANSITION_DURATION);

    return () => {
      clearTimeout(finishHidingMenu);
    };
  }, [visibleOnMobile]);

  const onHideMobile = () => {
    setVisibleOnMobile(false);

    if (mobileNavToggleRef) {
      // When menu is closed with escape key bring focus back to open close toggle
      mobileNavToggleRef.current?.focus();
    }
  };

  // Render modules
  const renderedModules = modules.map((module, index) =>
    module.render({
      key: index,
      slim,
      expandingOrCollapsing,
      onHideMobile,
      onSearchClick,
      currentPath,
      navigate,
    }),
  );

  return (
    <>
      <button
        onClick={onClickOpenCloseToggle}
        aria-label={gettext('Toggle sidebar')}
        aria-expanded={visibleOnMobile ? 'true' : 'false'}
        className={
          'button sidebar-nav-toggle' +
          (isMobile ? ' sidebar-nav-toggle--mobile' : '') +
          (visibleOnMobile ? ' sidebar-nav-toggle--open' : '')
        }
        type="button"
        ref={mobileNavToggleRef}
      >
        {visibleOnMobile ? <Icon name="cross" /> : <Icon name="bars" />}
      </button>
      <div
        className={
          'sidebar' +
          (slim ? ' sidebar--slim' : '') +
          (isMobile ? ' sidebar--mobile' : '') +
          (isMobile && !visibleOnMobile ? ' sidebar--hidden' : '') +
          (isMobile && !visibleOnMobile && closedOnMobile
            ? ' sidebar--closed'
            : '')
        }
      >
        <div
          className="sidebar__inner"
          onFocus={onFocusHandler}
          onBlur={onBlurHandler}
        >
          <div
            className={`sm:w-mt-2 ${
              slim ? 'w-justify-center' : 'w-justify-end'
            } w-flex  w-items-center`}
          >
            <button
              onClick={onClickCollapseToggle}
              aria-label={gettext('Toggle sidebar')}
              aria-expanded={slim ? 'false' : 'true'}
              type="button"
              className={`${!slim ? 'w-mr-4' : ''}
                button
                sidebar__collapse-toggle
                w-flex
                w-justify-center
                w-items-center
                hover:w-bg-surface-menu-item-active
                hover:text-white
                hover:opacity-100`}
            >
              <Icon
                name="expand-right"
                className={!collapsed ? '-w-rotate-180' : ''}
              />
            </button>
          </div>

          {renderedModules}
        </div>
      </div>
    </>
  );
};
