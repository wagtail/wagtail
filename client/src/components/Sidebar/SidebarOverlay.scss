.sidebar-overlay {
    z-index: 1100;
    width: 100vw;
    position: fixed;
    inset: 0;
    visibility: hidden;

    &--open {
        visibility: visible;
    }

    &__close-button {
        padding: 1em;
        color: #a5a5a5;
        border-bottom: 1px solid rgba(200, 200, 200, 0.1);
        cursor: pointer;
        display: block;
        background-color: #333;
        text-align: center;

        // Overrides for default link hover.
        &:hover {
            color: #a5a5a5;
            background-color: #222;
        }
    }
}
