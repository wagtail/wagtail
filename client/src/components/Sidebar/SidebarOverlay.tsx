/* eslint-disable react/prop-types */

import * as React from 'react';
import FocusTrap from 'focus-trap-react';

import Button from '../Button/Button';


export interface SidebarOverlayProps {
  isVisible: boolean;
  isOpen: boolean;
  onClose();
}

export const SidebarOverlay: React.FunctionComponent<SidebarOverlayProps> = (
  { isVisible, isOpen, onClose, children }) => {
  const className = (
    'sidebar-overlay'
    + (isVisible ? ' sidebar-overlay--visible' : '')
    + (isOpen ? ' sidebar-overlay--open' : '')
  );

  return (
    <FocusTrap
      paused={!isOpen}
      focusTrapOptions={{
        onDeactivate: onClose,
        fallbackFocus: '.sidebar-overlay__close-button'
      }}
    >
      <div role="dialog" className={className}>
        <Button onClick={onClose} className="sidebar-overlay__close-button">
          Close
        </Button>
        {children}
      </div>
    </FocusTrap>
  );
};
