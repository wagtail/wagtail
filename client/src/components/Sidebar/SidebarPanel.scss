.sidebar-panel {
  // With CSS variable allows panels with different widths to animate properly
  --width: #{$menu-width};

  visibility: hidden;
  transform: translate3d(0, 0, 0);
  position: fixed;
  height: 100vh;
  padding: 0;
  top: 0;
  // Remove once we drop support for Safari 13.
  // stylelint-disable-next-line property-disallowed-list
  left: 0;
  inset-inline-start: 0;
  z-index: 400;
  display: flex;
  flex-direction: column;
  overflow: hidden;

  @include transition(
    // Remove once we drop support for Safari 13.
    // stylelint-disable-next-line property-disallowed-list
    left $menu-transition-duration ease,
    inset-inline-start $menu-transition-duration ease
  );

  @include media-breakpoint-up(sm) {
    z-index: var(--z-index);
    width: var(--width);
    // Remove once we drop support for Safari 13.
    // stylelint-disable-next-line property-disallowed-list
    left: calc(#{$menu-width} - var(--width));
    inset-inline-start: calc(#{$menu-width} - var(--width));
  }

  @media (forced-colors: $media-forced-colours) {
    border-inline-start: 1px solid transparent;
    border-inline-end: 1px solid transparent;
  }

  &--visible {
    visibility: visible;
    box-shadow: 2px 0 2px rgba(0, 0, 0, 0.35);
  }

  @include media-breakpoint-up(sm) {
    @at-root .sidebar--slim #{&} {
      // Remove once we drop support for Safari 13.
      // stylelint-disable-next-line property-disallowed-list
      left: calc(#{$menu-width-slim} - var(--width));
      inset-inline-start: calc(#{$menu-width-slim} - var(--width));
    }
    // Don't apply this to nested submenus though
    @at-root .sidebar--slim .sidebar-panel #{&} {
      // Remove once we drop support for Safari 13.
      // stylelint-disable-next-line property-disallowed-list
      left: 0;
      inset-inline-start: 0;
    }

    &--open {
      // Remove once we drop support for Safari 13.
      // stylelint-disable-next-line property-disallowed-list
      left: $menu-width;
      inset-inline-start: $menu-width;

      // Don't apply this to nested submenus though
      @at-root .sidebar--slim .sidebar-panel #{&} {
        // Remove once we drop support for Safari 13.
        // stylelint-disable-next-line property-disallowed-list
        left: $menu-width;
        inset-inline-start: $menu-width;
      }
    }
  }
}
