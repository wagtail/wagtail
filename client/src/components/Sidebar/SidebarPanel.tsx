 

import * as React from 'react';

export interface SidebarPanelProps {
  isVisible: boolean;
  isOpen: boolean;
  depth: number;
  widthPx?: number;
}

export const SidebarPanel: React.FunctionComponent<SidebarPanelProps> = (
  { isVisible, isOpen, depth, widthPx, children }) => {
  const className = (
    'sidebar-panel'
    + (isVisible ? ' sidebar-panel--visible' : '')
    + (isOpen ? ' sidebar-panel--open' : '')
  );

  let zIndex = -depth * 2;

  const isClosing = isVisible && !isOpen;
  if (isClosing) {
    // When closing, make sure this panel displays behind any new panel that is opening
    zIndex--;
  }

  const style = {
    // See https://github.com/frenic/csstype#what-should-i-do-when-i-get-type-errors.
    ['--z-index' as any]: zIndex,
  };

  if (widthPx) {
    style['--width'] = widthPx + 'px';
  }

  return (
    <div className={className} style={style}>
      {children}
    </div>
  );
};
