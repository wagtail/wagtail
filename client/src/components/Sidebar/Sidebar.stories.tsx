import * as React from 'react';

import { ModuleDefinition, Sidebar, Strings } from './Sidebar';
import { WagtailBrandingModuleDefinition } from './modules/WagtailBranding';
import { SearchModuleDefinition } from './modules/Search';
import { MainMenuModuleDefinition } from './modules/MainMenu';
import { PageExplorerMenuItemDefinition } from './menu/PageExplorerMenuItem';
import { LinkMenuItemDefinition } from './menu/LinkMenuItem';
import { SubMenuItemDefinition } from './menu/SubMenuItem';
import { initFocusOutline } from '../../utils/focus';

import '../../../../wagtail/admin/static/wagtailadmin/css/sidebar.css';

export default { title: 'Sidebar/Sidebar', parameters: { layout: 'fullscreen' } };

const STRINGS: Strings = {
  DASHBOARD: 'Dashboard',
  EDIT_YOUR_ACCOUNT: 'Edit your account',
  SEARCH: 'Search',
};

function wagtailBrandingModule(): WagtailBrandingModuleDefinition {
  return new WagtailBrandingModuleDefinition('/admin/', {
    mobileLogo: 'https://wagtail.io/static/wagtailadmin/images/wagtail-logo.svg',
    desktopLogoBody: 'https://wagtail.io/static/wagtailadmin/images/logo-body.svg',
    desktopLogoTail: 'https://wagtail.io/static/wagtailadmin/images/logo-tail.svg',
    desktopLogoEyeOpen: 'https://wagtail.io/static/wagtailadmin/images/logo-eyeopen.svg',
    desktopLogoEyeClosed: 'https://wagtail.io/static/wagtailadmin/images/logo-eyeclosed.svg'
  });
}

function searchModule(): SearchModuleDefinition {
  return new SearchModuleDefinition('/admin/search/');
}

function bogStandardMenuModule(): MainMenuModuleDefinition {
  return new MainMenuModuleDefinition(
    [
      new PageExplorerMenuItemDefinition({
        name: 'explorer',
        label: 'Pages',
        url: '/admin/pages',
        icon_name: 'folder-open-inverse',
        classnames: '',
      }, 1),
      new LinkMenuItemDefinition({
        name: 'images',
        label: 'Images',
        url: '/admin/images/',
        icon_name: 'image',
        classnames: '',
      }),
      new LinkMenuItemDefinition({
        name: 'documents',
        label: 'Documents',
        url: '/admin/documents/',
        icon_name: 'doc-full-inverse',
        classnames: '',
      }),
      new LinkMenuItemDefinition({
        name: 'snippets',
        label: 'Snippets',
        url: '/admin/snippets/',
        icon_name: 'snippet',
        classnames: '',
      }),
      new LinkMenuItemDefinition({
        name: 'forms',
        label: 'Forms',
        url: '/admin/forms/',
        icon_name: 'form',
        classnames: '',
      }),
      new SubMenuItemDefinition(
        {
          name: 'reports',
          label: 'Reports',
          icon_name: 'site',
          classnames: '',
        },
        [
          new LinkMenuItemDefinition({
            name: 'locked-pages',
            label: 'Locked Pages',
            url: '/admin/reports/locked/',
            icon_name: 'lock',
            classnames: '',
          }),
          new LinkMenuItemDefinition({
            name: 'workflows',
            label: 'Workflows',
            url: '/admin/reports/workflow/',
            icon_name: 'tasks',
            classnames: '',
          }),
          new LinkMenuItemDefinition({
            name: 'workflow-tasks',
            label: 'Workflow tasks',
            url: '/admin/reports/workflow_tasks/',
            icon_name: 'thumbtack',
            classnames: '',
          }),
          new LinkMenuItemDefinition({
            name: 'site-history',
            label: 'Site history',
            url: '/admin/reports/site-history/',
            icon_name: 'history',
            classnames: '',
          }),
        ]
      ),
      new SubMenuItemDefinition(
        {
          name: 'settings',
          label: 'Settings',
          icon_name: 'cogs',
          classnames: '',
          footer_text: 'Wagtail Version',
        },
        [
          new LinkMenuItemDefinition({
            name: 'workflows',
            label: 'Workflows',
            url: '/admin/workflows/list/',
            icon_name: 'tasks',
            classnames: '',
          }),
          new LinkMenuItemDefinition({
            name: 'workflow-tasks',
            label: 'Workflow tasks',
            url: '/admin/workflows/tasks/index/',
            icon_name: 'thumbtack',
            classnames: '',
          }),
          new LinkMenuItemDefinition({
            name: 'users',
            label: 'Users',
            url: '/admin/users/',
            icon_name: 'user',
            classnames: '',
          }),
          new LinkMenuItemDefinition({
            name: 'groups',
            label: 'Groups',
            url: '/admin/groups/',
            icon_name: 'group',
            classnames: '',
          }),
          new LinkMenuItemDefinition({
            name: 'sites',
            label: 'Sites',
            url: '/admin/sites/',
            icon_name: 'site',
            classnames: '',
          }),
          new LinkMenuItemDefinition({
            name: 'collections',
            label: 'Collections',
            url: '/admin/collections/',
            icon_name: 'folder-open-1',
            classnames: '',
          }),
          new LinkMenuItemDefinition({
            name: 'redirects',
            label: 'Redirects',
            url: '/admin/redirects/',
            icon_name: 'redirect',
            classnames: '',
          }),
        ]
      ),
    ],
    [
      new LinkMenuItemDefinition({
        name: 'account',
        label: 'Account',
        url: '/admin/account/',
        icon_name: 'user',
        classnames: '',
      }),
      new LinkMenuItemDefinition({
        name: 'logout',
        label: 'Log out',
        url: '/admin/logout/',
        icon_name: 'logout',
        classnames: '',
      }),
    ],
    {
      name: 'Admin',
      avatarUrl: 'https://gravatar.com/avatar/e31ec811942afbf7b9ce0ac5affe426f?s=200&d=robohash&r=x',
    }
  );
}

interface RenderSidebarStoryOptions {
  rtl?: boolean;
  strings?: Strings;
}

function renderSidebarStory(
  modules: ModuleDefinition[], { rtl = false, strings = null }: RenderSidebarStoryOptions = {}
) {
  // Enable focus outlines so we can test them
  React.useEffect(() => {
    initFocusOutline();
  }, []);

  // Simulate navigation
  const [currentPath, setCurrentPath] = React.useState('/admin/');

  const navigate = (url: string) => {
    setCurrentPath(url);

    // Return resolved promise to close menu immediately
    return Promise.resolve();
  };

  // Add ready class to body to enable CSS transitions
  document.body.classList.add('ready');

  const onExpandCollapse = (collapsed: boolean) => {
    if (collapsed) {
      document.body.classList.add('sidebar-collapsed');
    } else {
      document.body.classList.remove('sidebar-collapsed');
    }
  };

  if (rtl) {
    document.documentElement.setAttribute('dir', 'rtl');
  } else {
    document.documentElement.setAttribute('dir', 'ltr');
  }

  return (
    <div className="wrapper">
      <Sidebar
        collapsedOnLoad={false}
        modules={modules}
        currentPath={currentPath}
        strings={strings || STRINGS}
        navigate={navigate}
        onExpandCollapse={onExpandCollapse}
      />
      <main id="main" className="content-wrapper" role="main">
        <div className="content">
          <b>Current path:</b> {currentPath}
        </div>
      </main>
    </div>
  );
}

export function standard() {
  return renderSidebarStory([
    wagtailBrandingModule(),
    searchModule(),
    bogStandardMenuModule(),
  ]);
}

export function withNestedSubmenu() {
  const menuModule = bogStandardMenuModule();

  menuModule.menuItems.push(
    new SubMenuItemDefinition(
      {
        name: 'nested-menu',
        label: 'Nested menu',
        icon_name: 'cogs',
        classnames: '',
      },
      [
        new LinkMenuItemDefinition({
          name: 'item',
          label: 'Item',
          url: '/admin/item/',
          icon_name: 'user',
          classnames: '',
        }),
        new SubMenuItemDefinition(
          {
            name: 'nested-menu',
            label: 'Nested menu',
            icon_name: 'folder-open-1',
            classnames: '',
          },
          [
            new LinkMenuItemDefinition({
              name: 'item',
              label: 'Item',
              url: '/admin/item/item/',
              icon_name: 'user',
              classnames: '',
            }),
            new SubMenuItemDefinition(
              {
                name: 'deeply-nested-menu',
                label: 'Deeply nested menu',
                icon_name: 'side',
                classnames: '',
              },
              [
                new LinkMenuItemDefinition({
                  name: 'item',
                  label: 'Item',
                  url: '/admin/item/item/item/',
                  icon_name: 'user',
                  classnames: '',
                }),
              ]
            ),
            new SubMenuItemDefinition(
              {
                name: 'another-deeply-nested-menu',
                label: 'Another deeply nested menu',
                icon_name: 'user',
                classnames: '',
              },
              [
                new LinkMenuItemDefinition({
                  name: 'item',
                  label: 'Item',
                  url: '/admin/item/item/item2/',
                  icon_name: 'user',
                  classnames: '',
                }),
              ]
            )
          ]
        )
      ]
    )
  );

  return renderSidebarStory([
    wagtailBrandingModule(),
    searchModule(),
    menuModule,
  ]);
}


export function withLargeSubmenu() {
  const menuModule = bogStandardMenuModule();

  const menuItems = [];
  for (let i = 0; i < 100; i++) {
    menuItems.push(
      new LinkMenuItemDefinition({
        name: `item-${i}`,
        label: `Item ${i}`,
        url: `/admin/item-${i}/`,
        icon_name: 'snippet',
        classnames: '',
      }),
    );
  }

  menuModule.menuItems.push(
    new SubMenuItemDefinition(
      {
        name: 'large-menu',
        label: 'Large menu',
        icon_name: 'cogs',
        classnames: '',
        footer_text: 'Footer text',
      },
      menuItems
    )
  );

  return renderSidebarStory([
    wagtailBrandingModule(),
    searchModule(),
    menuModule,
  ]);
}

export function withoutSearch() {
  return renderSidebarStory([
    wagtailBrandingModule(),
    bogStandardMenuModule(),
  ]);
}

// Translations taken from actual translation files at the time the code was written
// There were a few strings missing in reports/workflows. I left these as English as
// it's likely there will be a few untranslated strings on an Arabic site anyway.
const STRINGS_AR: Strings = {
  DASHBOARD: 'لوحة التحكم',
  EDIT_YOUR_ACCOUNT: 'تعديل حسابك',
  SEARCH: 'بحث',
};

function arabicMenuModule(): MainMenuModuleDefinition {
  return new MainMenuModuleDefinition(
    [
      new PageExplorerMenuItemDefinition({
        name: 'explorer',
        label: 'صفحات',
        url: '/admin/pages',
        start_page_id: 1,
        icon_name: 'folder-open-inverse',
        classnames: '',
      }),
      new LinkMenuItemDefinition({
        name: 'images',
        label: 'صور',
        url: '/admin/images/',
        icon_name: 'image',
        classnames: '',
      }),
      new LinkMenuItemDefinition({
        name: 'documents',
        label: 'وثائق',
        url: '/admin/documents/',
        icon_name: 'doc-full-inverse',
        classnames: '',
      }),
      new LinkMenuItemDefinition({
        name: 'snippets',
        label: 'قصاصات',
        url: '/admin/snippets/',
        icon_name: 'snippet',
        classnames: '',
      }),
      new LinkMenuItemDefinition({
        name: 'forms',
        label: 'نماذج',
        url: '/admin/forms/',
        icon_name: 'form',
        classnames: '',
      }),
      new SubMenuItemDefinition(
        {
          name: 'reports',
          label: 'التقارير',
          icon_name: 'site',
          classnames: '',
        },
        [
          new LinkMenuItemDefinition({
            name: 'locked-pages',
            label: 'Locked Pages',
            url: '/admin/reports/locked/',
            icon_name: 'lock',
            classnames: '',
          }),
          new LinkMenuItemDefinition({
            name: 'workflows',
            label: 'Workflows',
            url: '/admin/reports/workflow/',
            icon_name: 'tasks',
            classnames: '',
          }),
          new LinkMenuItemDefinition({
            name: 'workflow-tasks',
            label: 'Workflow tasks',
            url: '/admin/reports/workflow_tasks/',
            icon_name: 'thumbtack',
            classnames: '',
          }),
          new LinkMenuItemDefinition({
            name: 'site-history',
            label: 'Site history',
            url: '/admin/reports/site-history/',
            icon_name: 'history',
            classnames: '',
          }),
        ]
      ),
      new SubMenuItemDefinition(
        {
          name: 'settings',
          label: 'إعدادات',
          icon_name: 'cogs',
          classnames: '',
        },
        [
          new LinkMenuItemDefinition({
            name: 'workflows',
            label: 'Workflows',
            url: '/admin/workflows/list/',
            icon_name: 'tasks',
            classnames: '',
          }),
          new LinkMenuItemDefinition({
            name: 'workflow-tasks',
            label: 'Workflow tasks',
            url: '/admin/workflows/tasks/index/',
            icon_name: 'thumbtack',
            classnames: '',
          }),
          new LinkMenuItemDefinition({
            name: 'users',
            label: 'مستخدمين',
            url: '/admin/users/',
            icon_name: 'user',
            classnames: '',
          }),
          new LinkMenuItemDefinition({
            name: 'groups',
            label: 'مجموعات',
            url: '/admin/groups/',
            icon_name: 'group',
            classnames: '',
          }),
          new LinkMenuItemDefinition({
            name: 'sites',
            label: 'مواقع',
            url: '/admin/sites/',
            icon_name: 'site',
            classnames: '',
          }),
          new LinkMenuItemDefinition({
            name: 'collections',
            label: 'مجموعات',
            url: '/admin/collections/',
            icon_name: 'folder-open-1',
            classnames: '',
          }),
          new LinkMenuItemDefinition({
            name: 'redirects',
            label: 'اعادة التوجيهات',
            url: '/admin/redirects/',
            icon_name: 'redirect',
            classnames: '',
          }),
        ]
      ),
    ],
    [
      new LinkMenuItemDefinition({
        name: 'account',
        label: 'حساب',
        url: '/admin/account/',
        icon_name: 'user',
        classnames: '',
      }),
      new LinkMenuItemDefinition({
        name: 'logout',
        label: 'تسجيل الخروج',
        url: '/admin/logout/',
        icon_name: 'logout',
        classnames: '',
      }),
    ],
    {
      name: 'Admin',
      avatarUrl: 'https://gravatar.com/avatar/e31ec811942afbf7b9ce0ac5affe426f?s=200&d=robohash&r=x',
    }
  );
}

export function rightToLeft() {
  return renderSidebarStory([
    wagtailBrandingModule(),
    searchModule(),
    arabicMenuModule(),
  ], { rtl: true, strings: STRINGS_AR });
}
