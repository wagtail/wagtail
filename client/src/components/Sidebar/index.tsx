import * as React from 'react';
import ReactDOM from 'react-dom';
import Cookies from 'js-cookie';

import { Sidebar } from './Sidebar';

export const SIDEBAR_COLLAPSED_COOKIE_NAME = 'wagtail_sidebar_collapsed';

export function initSidebar() {
  const element = document.getElementById('wagtail-sidebar');
  const rawProps = document.getElementById('wagtail-sidebar-props');

  const navigate = (url: string) => {
    window.location.href = url;

    // Return a promise that never resolves.
    // This promise is used to indicate to any open submenus that the next page has loaded and it should close.
    // As all navigation from the menu at the moment takes the user to another page, we don't need to close the menus.
    // We will need to update this if we later add the ability to render views on the client side.
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    return new Promise<void>(() => {});
  };

  if (element && rawProps?.textContent) {
    const props = window.telepath.unpack(JSON.parse(rawProps.textContent));

    const collapsedCookie: any = Cookies.get(SIDEBAR_COLLAPSED_COOKIE_NAME);
    // Cast to boolean
    const collapsed = !((collapsedCookie === undefined || collapsedCookie === '0'));

    const onExpandCollapse = (_collapsed: boolean) => {
      if (_collapsed) {
        document.body.classList.add('sidebar-collapsed');
        Cookies.set(SIDEBAR_COLLAPSED_COOKIE_NAME, 1);
      } else {
        document.body.classList.remove('sidebar-collapsed');
        Cookies.set(SIDEBAR_COLLAPSED_COOKIE_NAME, 0);
      }
    };

    ReactDOM.render(
      <Sidebar
        modules={props.modules}
        strings={wagtailConfig.STRINGS}
        collapsedOnLoad={collapsed}
        currentPath={window.location.pathname}
        navigate={navigate}
        onExpandCollapse={onExpandCollapse}
      />,
      element,
      () => {
        document.body.classList.add('ready');
      }
    );
  }
}
