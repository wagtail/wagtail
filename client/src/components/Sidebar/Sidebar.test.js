import React from 'react';
import { shallow } from 'enzyme';
import { Sidebar } from './Sidebar';

describe('Sidebar', () => {
  const strings = {};

  it('should render with the minimum required props', () => {
    const wrapper = shallow(<Sidebar modules={[]} strings={strings} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should toggle the slim mode in the sidebar when outer button clicked', () => {
    const onExpandCollapse = jest.fn();

    const wrapper = shallow(
      <Sidebar
        modules={[]}
        onExpandCollapse={onExpandCollapse}
        strings={strings}
      />,
    );

    // default expanded (non-slim)
    expect(
      wrapper.find('.sidebar__collapse-toggle').prop('aria-expanded'),
    ).toEqual('true');
    expect(wrapper.find('.sidebar--slim')).toHaveLength(0);
    expect(onExpandCollapse).not.toHaveBeenCalled();

    // toggle slim mode
    wrapper.find('.sidebar__collapse-toggle').simulate('click');

    expect(
      wrapper.find('.sidebar__collapse-toggle').prop('aria-expanded'),
    ).toEqual('false');
    expect(wrapper.find('.sidebar--slim')).toHaveLength(1);
    expect(onExpandCollapse).toHaveBeenCalledWith(true);
  });

  it('should toggle the sidebar visibility on click (used on mobile)', () => {
    const onExpandCollapse = jest.fn();

    const wrapper = shallow(
      <Sidebar
        modules={[]}
        onExpandCollapse={onExpandCollapse}
        strings={strings}
      />,
    );

    // default not expanded
    expect(wrapper.find('.sidebar-nav-toggle').prop('aria-expanded')).toEqual(
      'false',
    );
    expect(wrapper.find('.sidebar-nav-toggle--open')).toHaveLength(0);

    // toggle expanded mode
    wrapper.find('.sidebar-nav-toggle').simulate('click');

    // check it is expanded
    expect(wrapper.find('.sidebar-nav-toggle').prop('aria-expanded')).toEqual(
      'true',
    );
    expect(wrapper.find('.sidebar-nav-toggle--open')).toHaveLength(1);
  });
});
