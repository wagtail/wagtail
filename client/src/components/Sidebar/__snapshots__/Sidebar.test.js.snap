// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar should render with the minimum required props 1`] = `
<Fragment>
  <div
    className="sidebar"
  >
    <div
      className="sidebar__inner"
    >
      <button
        aria-expanded="true"
        className="button sidebar__collapse-toggle hover:w-bg-primary-200 hover:text-white hover:opacity-100"
        onClick={[Function]}
        type="button"
      >
        <Icon
          className="w-transition motion-reduce:w-transition-none
                -w-rotate-180
                "
          name="expand-right"
        />
      </button>
      <div
        className="sidebar__peek-hover-area"
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      />
    </div>
  </div>
  <button
    aria-expanded="false"
    className="button sidebar-nav-toggle"
    onClick={[Function]}
    type="button"
  >
    <Icon
      name="bars"
    />
  </button>
</Fragment>
`;
